include/revng/TupleTree/TrackingImpl.h

```diff
@@ -80,8 +80,13 @@ struct TrackingImpl {
 
   template<typename M, StrictSpecializationOf<UpcastablePointer> T>
   static void collectImpl(const T &UP, TupleTreePath &Stack, ReadFields &Info) {
-    if (!UP.empty())
-      UP.upcast([&](auto &Upcasted) { collectImpl<M>(Upcasted, Stack, Info); });
+    if (!UP.empty()) {
+      UP.upcast([&](auto &Upcasted) {
+        Stack.push_back(Upcasted.Kind());
+        collectImpl<M>(Upcasted, Stack, Info);
+        Stack.pop_back();
```

Comment:
Context?

Reply:
After a lengthy mattermost discussion we ended up with a can conclusion that with can't get away without adding the kind of a polymorphic type to the path. This is a consequence of that.

Reply:
Ah, OK, I thought this was something completely different.
Can you add a comment to clarify?

Reply:
Done!